### PR TITLE
Fix: MinerU parsed images are obscured when using MinerU as a parser.…

### LINKED ISSUE #14197: [Bug]: Why are the parsed images obscured when I use MinerU as a parser?
### Self Checks

- [x] I have searched for existing issues [search for existing issues](https://github.com/infiniflow/ragflow/issues), including closed ones.
- [x] I confirm that I am using English to submit this report ([Language Policy](https://github.com/infiniflow/ragflow/issues/5910)).
- [x] Non-english title submitions will be closed directly ( 非英文标题的提交将会被直接关闭 ) ([Language Policy](https://github.com/infiniflow/ragflow/issues/5910)).
- [x] Please do not modify this template :) and fill in all the required fields.

### RAGFlow workspace code commit ID

0.24.0

### RAGFlow image version

0.24.0

### Other environment information

```Markdown

```

### Actual behavior

Why are the parsed images obscured when I use MinerU as a parser? , as follows:

Image

### Expected behavior

_No response_

### Steps to reproduce

```Markdown
There should be no obstruction.
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Lobster-0429**
### What problem does this PR solve?

When using MinerU as a document parser, parsed images appeared obscured (darkened) in the RAGFlow UI. The root cause was in the `crop` method of `deepdoc/parser/mineru_parser.py`: a semi-transparent black overlay (intended only for surrounding context strips) was incorrectly applied to actual content images when the top context strip crop was skipped due to zero-height regions.

The fix tracks the `is_context` flag alongside each crop at generation time (`imgs_with_flags`), ensuring the overlay is only applied to context strips — never to the actual parsed content.

Fixes #14197

### Type of change

- [x] Bug Fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by web-dev0521** [comment]
Hi @yingfeng,
This is my first contribution to the infiniflow/ragflow repository. I’ve put significant effort into this PR and aimed to follow the project’s conventions and standards.
I would appreciate your review whenever you have time, and I welcome any feedback or suggestions for improvement.
Thank you.

File: deepdoc/parser/mineru_parser.py [hunk 1]
```diff
@@ -416,7 +416,11 @@ def crop(self, text, ZM=1, need_position=False):
         )
 
         positions = []
+        # Each entry is (PIL.Image, is_context) so overlay logic is decoupled from final index.
+        imgs_with_flags = []
+        poss_last_idx = len(poss) - 1
```
[the remark this answers is not in the document]
Comment (Lobster-0429) [reply deepdoc/parser/mineru_parser.py]: Good catch — thank you for pointing that out.

The imgs = [] line has already been removed in the latest commit. The unused variable was identified and cleaned up to address the Ruff F841 warning.

CI should now pass cleanly, but please let me know if anything else needs attention.

**Comment by web-dev0521** [comment]
Hi @yingfeng ,
Hope you had a great weekend.
I’ve addressed all the issues raised. Thank you very much for your time and valuable feedback—I truly appreciate it.
Please let me know if there’s anything else I should update.

**Comment by web-dev0521** [comment]
Hi @yingfeng, thank you for your attention and guidance.
Hi @xugangqiang, could you please take a moment to review my PR when convenient? I would greatly appreciate your feedback.
Thank you both for your time.

**Review by 6ba3i — COMMENTED** [review]
Thanks for the fix here. I think this PR is addressing the right bug in the right place.
That said, I think this can be made smaller.

The current PR keeps extra state for every cropped image with `imgs_with_flags`, and also adds debug logging and some helper/comment churn. For this specific bug, I do not think we need all of that. A smaller fix is enough:

- keep `imgs` as plain images
- record where the leading synthetic context segment ends
- record where the trailing synthetic context segment begins
- use those two boundaries in the final overlay loop instead of `ii == 0` / `ii + 1 == len(imgs)`

Suggested change:

```diff
diff --git a/deepdoc/parser/mineru_parser.py b/deepdoc/parser/mineru_parser.py
@@
-        positions = []
+        positions = []
+        head_ctx_end = 0
+        tail_ctx_start = 0
         for ii, (pns, left, right, top, bottom) in enumerate(poss):
+            if ii + 1 == len(poss):
+                tail_ctx_start = len(imgs)
             right = left + max_width
@@
                 cimgp = page.crop((x0, y0, x1, y1))
                 imgs.append(cimgp)
                 if 0 < ii < len(poss) - 1:
                     positions.append((pn + self.page_from, x0, x1, y0, y1))
                 bottom -= page.size[1]
+            if ii == 0:
+                head_ctx_end = len(imgs)
@@
         height = 0
         for ii, img in enumerate(imgs):
-            if ii == 0 or ii + 1 == len(imgs):
+            if ii < head_ctx_end or ii >= tail_ctx_start:
                 img = img.convert("RGBA")
                 overlay = Image.new("RGBA", img.size, (0, 0, 0, 0))
                 overlay.putalpha(128)
```
it fixes the same regression with less churn and less code change, could you give it a try ? thanks again !

**Comment by web-dev0521** [comment]
Hi, @6ba3i !
Thank you for your attention.
I'll update it asap.

**Comment by xugangqiang** [comment]
@web-dev0521 
May I know if proper testing has been conducted, like regrssion test?
If so, please uploading testing evidence so that we know it's tested and will not cause regression issues.

1. what's the UI (screen shots) before fix
2. what's after fix (screen shots, logs, ..., etc)

Appreciate your contributions and looking forward for more contributions.

**Comment by web-dev0521** [comment]
Hi @6ba3i,
I’ve updated the code based on your recommendations. I would appreciate it if you could take a moment to review the changes when convenient.
Thank you for your time and guidance.

**Review by 6ba3i — COMMENTED** [review]
the code looks good to me @web-dev0521 ! do you mind answering back @xugangqiang on his requests so that i can formally approve this pr ?

**Comment by web-dev0521** [comment]
Hi, @xugangqiang !
Thank you for the thorough review and for the helpful suggestion regarding the boundary-index approach - I have incorporated it as recommended.

To ensure accuracy and provide proper regression coverage, I’ve added unit tests in `test/unit_test/deepdoc/parser/test_mineru_crop.py` that reproduce the reported edge cases.

The tests cover:

* Content image at `top = 0` (zero-height top context strip, which is skipped)
* Content image within the GAP of the page top
* The core scenario identified in the discussion where both context strips are skipped simultaneously (`len(imgs) == 1`)
* Verification that context strips are still correctly darkened (ensuring no regression)

All four tests are passing and confirm that the fix behaves as expected. These tests should also provide reliable regression coverage in CI to prevent similar issues in the future.

Thank you again for your guidance and support. I look forward to your feedback.

File: test/unit_test/deepdoc/parser/test_mineru_crop.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+Regression tests for issue #14197: content images were incorrectly darkened
+when the top context strip had zero height and was skipped.
+"""
+
+import logging
+
+from PIL import Image
+
+from deepdoc.parser.mineru_parser import MinerUParser
```
[the remark this answers is not in the document]
Comment (Lobster-0429) [reply test/unit_test/deepdoc/parser/test_mineru_crop.py:28]: Added `@pytest.mark.p1` to the core regression test (`test_single_image_not_darkened_when_both_context_strips_skipped`) and `@pytest.mark.p2` to the remaining tests. The corresponding markers have been properly registered in `pyproject.toml` under `[tool.pytest.ini_options]`.

**Comment by web-dev0521** [comment]
Hello, @yingfeng!
I have completed all the actions regarding the points you pointed out.
Thank you so much for taking the time to provide feedback—and thanks to everyone else who offered their valuable input as well. I am truly grateful.
Please feel free to let me know anytime if there are any further updates I should make.

**Comment by yingfeng** [comment]
> Hello, @yingfeng! I have completed all the actions regarding the points you pointed out. Thank you so much for taking the time to provide feedback—and thanks to everyone else who offered their valuable input as well. I am truly grateful. Please feel free to let me know anytime if there are any further updates I should make.

Can you answer back @xugangqiang on his requests ?
